src/registry/http.ts

```diff
@@ -0,0 +1,453 @@
+  const url = urlObject.toString();
+  const parts = authenticateHeader.split(" ");
+  if (parts.length === 0) {
+    throw new Error(`can't retrieve WWW-Authenticate header in /v2 endpoint on registry ${url}: malformed`);
+  }
+
+  const authType = parts[0].toLowerCase();
+  switch (authType) {
+    case "bearer":
+    case "basic":
+      break;
+    default:
+      throw new Error(`unsupported auth type in WWW-Authenticate on registry ${url}: ${parts[0]}`);
+  }
+
+  const variables = parts[1].split(",");
+  const authContextOptional: Partial<AuthContext> = {};
+  variables.forEach((variable) => {
+    const firstEqual = variable.indexOf("=");
+    if (firstEqual === -1) {
+      throw new Error(`expected '=' but didn't encounter it on Auth header on registry ${url}`);
+    }
+
+    const name = variable.slice(0, firstEqual);
+    const value = variable.slice(firstEqual + 1);
+    const isMalformed = value.length < 2 || value[0] !== `"` || value[value.length - 1] !== `"`;
+    if (isMalformed) {
+      throw new Error(`malformed value on auth header on registry ${url}`);
+    }
+
+    const trimmedValue = value.slice(1, value.length - 1);
+    switch (name) {
+      case "realm":
+      case "scope":
+      case "service":
+        authContextOptional[name] = trimmedValue;
+        break;
+      default:
+        console.debug(`unknown auth attribute ${name} on registry ${url}`);
+    }
+  });
+
+  if (!authContextOptional.realm) throw new Error(`expected a realm on the auth header in repository ${url}`);
+  try {
+    const urlRealm = new URL(authContextOptional.realm);
+    // if service is not defined, define it by setting it to be the same as the realm's host.
+    // at the end service will be used in the request to know which service to hit
+    authContextOptional.service ??= urlRealm.host;
+  } catch {
+    throw new Error(`invalid url in realm in repository ${url}`);
+  }
+
+  return {
+    authType,
+    realm: authContextOptional.realm!,
+    scope: authContextOptional.scope ?? "",
+    service: authContextOptional.service ?? "",
+  };
+}
+
+// RegistryHTTPClient implements a registry client that is able to pull/push to the configured registry
+export class RegistryHTTPClient implements Registry {
+  private url: URL;
+
+  constructor(private env: Env, private configuration: RegistryConfiguration) {
+    this.url = new URL(configuration.registry);
+  }
+
+  authBase64(): string {
+    return btoa(this.configuration.username + ":" + this.password());
+  }
+
+  password(): string {
+    return (this.env as unknown as Record<string, string>)[this.configuration.password_env] ?? "";
+  }
+
+  async authenticate(): Promise<HTTPContext> {
+    const res = await fetch(`${this.url.protocol}//${this.url.host}/v2/`, {
+      headers: {
+        "User-Agent": "Docker-Client/24.0.5 (linux)",
+        "Accept-Encoding": "gzip",
+      },
+    });
+    if (res.status !== 401) {
```

Comment:
Can you handle the case when no authentication is required?

```
if (res.status === 200) {
  return {
    authContext: {
      authType: "none",
      realm: "",
      scope: "",
      service: this.url.host,
    },
    repository: "",
    accessToken: "",
  };
}
```

And add `none` to `AuthType`